tools/osv-linter/internal/pkgchecker/ecosystems.go

```diff
@@ -0,0 +1,340 @@
+	case "Rocky Linux":
+		return nil
+	case "RubyGems":
+		return nil
+	case "SwiftURL":
+		return nil
+	case "Ubuntu":
+		return nil
+	case "UVI":
+		return nil
+	case "Wolfi":
+		return nil
+	}
+	return fmt.Errorf("unsupported ecosystem: %s", ecosystem)
+}
+
+// Validate the existence of a package in PyPI.
+func existsInPyPI(pkg string) bool {
+	packageInstanceURL := fmt.Sprintf("https://pypi.org/pypi/%s/json", pkg)
+
+	// This 404's for non-existent packages.
+	resp, err := faulttolerant.Head(packageInstanceURL)
+	if err != nil {
+		return false
+	}
+
+	return resp.StatusCode == http.StatusOK
+}
+
+// Confirm that all specified versions of a package exist in PyPI.
+func versionsExistInPyPI(pkg string, versions []string) error {
+	packageInstanceURL := fmt.Sprintf("https://pypi.org/pypi/%s/json", pkg)
+
+	// This 404's for non-existent packages.
+	resp, err := faulttolerant.Get(packageInstanceURL)
+	if err != nil {
+		return fmt.Errorf("unable to validate package: %v", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unable to validate package: %q for %s", resp.Status, packageInstanceURL)
+	}
+
+	// Parse the known versions from the JSON.
+	respJSON, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve JSON for %q: %v", pkg, err)
+	}
+	// Fetch all known versions of package.
+	versionsInPyPy := []string{}
+	releases := gjson.GetBytes(respJSON, "releases.@keys")
+	releases.ForEach(func(key, value gjson.Result) bool {
+		versionsInPyPy = append(versionsInPyPy, value.String())
+		return true // keep iterating.
+	})
+	// Determine which referenced versions are missing.
+	versionsMissing := []string{}
+	for _, versionToCheckFor := range versions {
+		if slices.Contains(versionsInPyPy, versionToCheckFor) {
+			continue
+		}
+		versionsMissing = append(versionsMissing, versionToCheckFor)
+	}
+	if len(versionsMissing) > 0 {
+		return fmt.Errorf("failed to find %#v for %q in PyPI", versionsMissing, pkg)
+	}
+
+	return nil
+}
+
+// Validate the existence of a package in Go.
+func existsInGo(pkg string) bool {
+	// Of course the Go runtime exists :-)
+	if pkg == "stdlib" || pkg == "toolchain" {
+		return true
+	}
+
+	// The Go Module Proxy seems to require package names to be lowercase.
+	// GitHub URLs are known to be case-insensitive.
+	if strings.HasPrefix(pkg, "github.com/") {
+		pkg = strings.ToLower(pkg)
+	}
+
+	packageInstanceURL := fmt.Sprintf("https://proxy.golang.org/%s/@v/list", pkg)
+
+	// This 404's for non-existent packages.
+	resp, err := faulttolerant.Head(packageInstanceURL)
+	if err != nil {
+		return false
+	}
+	return resp.StatusCode == http.StatusOK
+}
+
+// isGoPseudoVersion checks if a given version string is a Go pseudo-version,
+// including those with pre-release and build metadata segments,
+// and handles cases where the pre-release identifier starts with '0.'.
+func isGoPseudoVersion(version string) bool {
```

Comment:
not sure if this would help? https://pkg.go.dev/golang.org/x/mod/module#IsPseudoVersion

Reply:
That's the sort of cool find I was hoping you might bring to my attention, thank you!

It doesn't seem to match on versions like `1.16.0-0`, which seems to be in use on records like https://storage.googleapis.com/go-vulndb/ID/GO-2021-0067.json and aren't in the set of known Go versions returned by https://go.dev/dl/?mode=json&include=all

Do you know of any particular way I should treat these versions? Otherwise I'll just use `strings.HasSuffix(v, "-0")`

Reply:
9fc61f5 switches things up, thank you

Reply:
I don't know anything particular about that :( did you sort it out?

Reply:
The current semi-gross solution:

https://github.com/ossf/osv-schema/blob/8907bae6ff34ccd0480b628f49077282772e7b4a/tools/osv-linter/internal/pkgchecker/ecosystems.go#L334-L343